src/coreclr/jit/valuenum.cpp

```diff
@@ -8774,6 +8788,54 @@ void Compiler::fgValueNumberTree(GenTree* tree)
         }
         else // Look up the VNFunc for the node
         {
+            // Check if we can fold GT_ARR_LENGTH on top of a known array (immutable)
+            if (tree->OperIs(GT_ARR_LENGTH))
+            {
+                // Case 1: ARR_LENGTH(FROZEN_OBJ)
+                ValueNum addressVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+                if (vnStore->IsVNHandle(addressVN) && (vnStore->GetHandleFlags(addressVN) == GTF_ICON_OBJ_HDL))
+                {
+                    size_t handle = vnStore->CoercedConstantValue<size_t>(addressVN);
+                    int    len    = this->info.compCompHnd->getArrayOrStringLength((CORINFO_OBJECT_HANDLE)handle);
+                    if (len >= 0)
+                    {
+                        tree->gtVNPair.SetBoth(vnStore->VNForIntCon(len));
+                        return;
+                    }
+                }
+
+                // Case 2: ARR_LENGTH(static-readonly-field)
+                VNFuncApp funcApp;
+                if (vnStore->GetVNFunc(addressVN, &funcApp) && (funcApp.m_func == VNF_InvariantNonNullLoad))
+                {
+                    ValueNum fieldSeqVN = funcApp.m_args[1];
+                    if ((fieldSeqVN != vnStore->VNForNull()) && (fieldSeqVN != ValueNumStore::NoVN))
+                    {
+                        FieldSeq* fieldSeq = vnStore->FieldSeqVNToFieldSeq(fieldSeqVN);
+                        if (fieldSeq != nullptr)
+                        {
+                            CORINFO_FIELD_HANDLE field = fieldSeq->GetFieldHandle();
+                            if (field != NULL)
+                            {
+                                uint8_t buffer[TARGET_POINTER_SIZE] = {0};
+                                if (this->info.compCompHnd->getReadonlyStaticFieldValue(field, buffer,
+                                                                                        TARGET_POINTER_SIZE, false))
+                                {
+                                    CORINFO_OBJECT_HANDLE objHandle;
+                                    memcpy(&objHandle, buffer, TARGET_POINTER_SIZE);
```

Comment:
This does not look right. `CORINFO_OBJECT_HANDLE` is not `TARGET_POINTER_SIZE`d, it is host pointer sized.

Reply:
thanks, fixed

Reply:
NAOT's implementation of `getReadonlyStaticFieldValue` looks questionable too. It is mixing `nint` and target pointer size concepts. I don't know if NAOT supports cross bitness compilation, though.

Reply:
@jakobbotsch it does support that, can you point me what exactly confuses you? I don't see any `nint` there

Reply:
btw, I assume it's never possible to have 32bit crosscompiling to 64bit on any of our targets

Reply:
> can you point me what exactly confuses you

This part:
https://github.com/dotnet/runtime/blob/3dbc850af3e8bfd6d529ed90cf00247dc9a24512/src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs#L2254-L2260

Reply:
@jakobbotsch it should be correct from my understanding, `handle` despite being `nint` is actually 32bit value in case of ILC-x64 targeting 32bit. cc @jkotas to confirm (who asked me to use target-pointer-size actually)

Reply:
Actually it probably was correct as is with TARGET_POINTER_SIZE on jit side 🤔

Reply:
reverted changes, should be good as is - in other places it expects `getReadonlyStaticFieldValue` to return 4-byte reference type on x64 bit jit targeting x86 `genTypeSize(TYP_REF) == 4`.

The frozen handles at the same time are expected to be always 32bit on that

Reply:
In that case `CORINFO_OBJECT_HANDLE` should be typedef for `target_size_t`. The contract about this should be clearer.

If R2R/NAOT guarantee that handles are target pointer size then `nint` is odd choice as well (btw, at the minimum it is losing `s_handleHighBitSet`, but `HandleToObject` does not verify anything about this bit surviving, so I guess it works out in practice?).

Reply:
> reverted changes, should be good as is - in other places it expects getReadonlyStaticFieldValue to return 4-byte reference type on x64 bit jit targeting x86 genTypeSize(TYP_REF) == 4.

The `memcpy` is leaving upper half of `objHandle` with stack garbage for 64-bit host targeting 32-bit target.

Reply:
> > reverted changes, should be good as is - in other places it expects getReadonlyStaticFieldValue to return 4-byte reference type on x64 bit jit targeting x86 genTypeSize(TYP_REF) == 4.
> 
> The `memcpy` is leaving upper half of `objHandle` with stack garbage for 64-bit host targeting 32-bit target.

fixed

Reply:
> If R2R/NAOT guarantee that handles are target pointer size then nint is odd choice as well 

I'd love to avoid rewritting half of NAOT as part of this PR 🙂

Reply:
> In that case CORINFO_OBJECT_HANDLE should be typedef for target_size_t.

Does it add value? we'll need to define these types in VM (target.h is a jit thing), crossgen and ILC

Reply:
The odd choice here is that these handles are target pointer sized instead of host pointer sized. Is it because JIT directly puts these handles into the code when referring to these frozen objects?

Reply:
In general, the JIT/EE interface is trying to distinguish between handles that are used to pass information on JIT/EE interface (e.g. `CORINFO_CLASS_HANDLE`) and how the underlying runtime value that should be embedded into the code (e.g. value returned by `embedClassHandle`).

We have been taking shortcut here and this shortcut does not seem to be aging well.

Reply:
@jkotas so is it actionable as part of this PR?

Reply:
Similar shortcut exists in a few other places in the JIT. I do not think it needs to be fixed as part of this PR. It is something to keep in mind if we run into more issues with it.

Reply:
@jakobbotsch can you please approve it if the rest is good then?